src/data/apint.rs

```diff
@@ -0,0 +1,382 @@
+            unsafe{ self.drop_digits(); }
+
+            match rhs.storage() {
+                Storage::Inl => {
+                    // If `rhs` is a small `ApInt` we can simply update
+                    // the `digit` field of `self` and we are done.
+                    self.data.inl = unsafe{ rhs.data.inl };
+                }
+                Storage::Ext => {
+                    // If `rhs` is a large heap-allocated `ApInt` we first
+                    // need to expensively clone its buffer and feed it to `self`.
+                    let cloned = rhs.clone();
+                    self.data.ext = unsafe{ cloned.data.ext };
+                    use std::mem;
+                    mem::forget(cloned);
+                }
+            }
+        }
+        // Since all cases may need bit width adjustment we outsourced it
+        // to the end of this method.
+        self.len = rhs.len;
+    }
+
+    /// Strictly assigns `rhs` to this `ApInt`.
+    /// 
+    /// After this operation `rhs` and `self` are equal to each other.
+    /// 
+    /// **Note:** Strict assigns protect against mutating the bit width
+    /// of `self` and thus return an error instead of executing a probably
+    /// expensive `assign` operation.
+    /// 
+    /// # Errors
+    /// 
+    /// - If `rhs` and `self` have unmatching bit widths.
+    pub fn strict_assign(&mut self, rhs: &ApInt) -> Result<()> {
+        if self.width() != rhs.width() {
+            return Error::unmatching_bitwidths(self.width(), rhs.width())
+                .with_annotation(format!(
+                    "Occured while trying to `strict_assign` {:?} to {:?}.", self, rhs))
+                .into()
+        }
+        self.as_digit_slice_mut()
+            .copy_from_slice(rhs.as_digit_slice());
+        Ok(())
+    }
+}
+
+impl Drop for ApInt {
+    fn drop(&mut self) {
+        unsafe{self.drop_digits()}
+    }
+}
+
+impl Clone for ApInt {
+    fn clone(&self) -> Self {
+        match self.storage() {
+            Storage::Inl => {
+                ApInt::new_inl(self.len, unsafe{ self.data.inl })
+            }
+            Storage::Ext => {
+                use std::mem;
+                let req_digits = self.len_digits();
+                //TODO: this can be simplified and should be benchmarked
+                let mut buffer = self.as_digit_slice()
+                    .to_vec()
+                    .into_boxed_slice();
+                assert_eq!(buffer.len(), req_digits);
+                let ptr_buffer = buffer.as_mut_ptr();
+                mem::forget(buffer);
+                unsafe{ ApInt::new_ext(self.len, ptr_buffer) }
+            }
+        }
+    }
+}
+
+impl Width for ApInt {
+    /// Returns the `BitWidth` of this `ApInt`.
+    #[inline]
+    fn width(&self) -> BitWidth {
+        BitWidth::new(self.len_bits()).unwrap()
+    }
+}
+
+//this is put here to be able to constrict the publicity of `ApIntData`
```

Comment:
The one problem we will have with the reorganization is that I could not make some of the fields of `Apint` and `ApIntData` private. I moved some test code (that used private fields) into that file so that I could constrict the visibility as much as I could. Everything except for the `len` field is constricted to `pub(in crate::data)`, and from there I searched the codebase manually to make sure the fields were only being directly accessed in the right places.